File: src/php/PhantasyIslandPhpRemoteControl/AirplaneManager.php
<?php

namespace PhantasyIslandPhpRemoteControl;

/**
 * 管理并更新所有飞机状态的管理器
 */
class AirplaneManager
{
    private static $instance = null;
    public $airplanes_table = [];

    public static function getInstance()
    {
        if (self::$instance === null) {
            self::$instance = new AirplaneManager();
        }
        return self::$instance;
    }

    public function ping()
    {
        return HttpLayer::ping();
    }

    public function ping_volatile()
    {
        return HttpLayer::ping_volatile();
    }

    public function start()
    {
        return HttpLayer::start();
    }

    public function start_volatile()
    {
        return HttpLayer::start_volatile();
    }

    public function get_airplane($id)
    {
        return $this->airplanes_table[$id] ?? null;
    }

    public function sleep($time)
    {
        if ($time < 1) {
            usleep($time * 1000000);
        } else {
            sleep($time);
        }
    }

    public function flush()
    {
        $status_data = HttpLayer::get_all_airplane_status();
        $airplane_status = HttpLayer::process_airplane($status_data);
        
        if ($airplane_status !== null) {
            foreach ($airplane_status as $k => $status) {
                if (!isset($this->airplanes_table[$k])) {
                    $this->airplanes_table[$k] = new AirplaneController(
                        $status['keyName'],
                        $status['typeName'],
                        $status['updateTimestamp'],
                        $status['status'],
                        $status['cameraFront'],
                        $status['cameraDown']
                    );
                } else {
                    $a = $this->airplanes_table[$k];
                    $a->keyName = $status['keyName'];
                    $a->typeName = $status['typeName'];
                    $a->updateTimestamp = $status['updateTimestamp'];
                    $a->status = new AirplaneFlyStatus($status['status']);
                    $a->cameraFront = $status['cameraFront'];
                    $a->cameraDown = $status['cameraDown'];
                }
            }
        }
        return $airplane_status;
    }
}

function get_airplane_manager()
{
    return AirplaneManager::getInstance();
}

File: src/rb/PhantasyIslandRubyRemoteControl/airplane_manager.rb
require_relative 'airplane_core'
require_relative 'control_command'
require_relative 'http_layer'

# 管理并更新所有飞机状态的管理器
class AirplaneManager
  attr_accessor :airplanes_table

  def initialize
    @airplanes_table = {}
  end

  # 测试与仿真平台的连接状态
  def ping
    HttpLayer.ping
  end

  def ping_volatile
    HttpLayer.ping_volatile
  end

  # 向仿真场景发出开始仿真指令
  def start
    HttpLayer.start
  end

  def start_volatile
    HttpLayer.start_volatile
  end

  # 获取指定无人机
  def get_airplane(id)
    @airplanes_table[id]
  end

  def sleep(time)
    Kernel.sleep(time)
  end

  # 更新当前管理器所管理的所有飞机的状态
  def flush
    airplane_status = HttpLayer.process_airplane(HttpLayer.get_all_airplane_status)
    if airplane_status
      airplane_status.each do |k, status|
        if @airplanes_table[k].nil?
          @airplanes_table[k] = AirplaneController.new(
            keyName: status['keyName'],
            typeName: status['typeName'],
            updateTimestamp: status['updateTimestamp'],
            status: AirplaneFlyStatus.from_h(status['status']),
            cameraFront: status['cameraFront'],
            cameraDown: status['cameraDown']
          )
        else
          a = @airplanes_table[k]
          a.keyName = status['keyName']
          a.typeName = status['typeName']
          a.updateTimestamp = status['updateTimestamp']
          a.status = AirplaneFlyStatus.from_h(status['status'])
          a.cameraFront = status['cameraFront']
          a.cameraDown = status['cameraDown']
        end
      end
    else
      nil
    end
  end
end

# 单例模式
$airplane_manager_singleton = AirplaneManager.new

# 获取AirplaneManager单例对象
def get_airplane_manager
  $airplane_manager_singleton
end
